src/imitation/regularization/updaters.py

```diff
@@ -0,0 +1,139 @@
+"""Implements parameter scaling algorithms to update the parameters of a regularizer."""
+
+from typing import Protocol, Tuple, Union
+
+import numpy as np
+import torch as th
+
+LossType = Union[th.Tensor, float]
+
+
+class LambdaUpdater(Protocol):
+    """Protocol type for functions that update the regularizer parameter.
+
+    A callable object that takes in the current lambda and the train and val loss, and
+    returns the new lambda. This has been implemented as a protocol and not an ABC
+    because a user might wish to provide their own implementation without having to
+    inherit from the base class, e.g. by defining a function instead of a class.
+
+    Note: if you implement `LambdaUpdater`, your implementation MUST be purely
+    functional, i.e. side-effect free. The class structure should only be used
+    to store constant hyperparameters. (Alternatively, closures can be used for that).
+    """
+
+    def __call__(self, lambda_, train_loss: LossType, val_loss: LossType) -> float:
+        ...
+
+
+class IntervalParamScaler(LambdaUpdater):
+    """Scales the lambda of the regularizer by some constant factor.
+
+    Lambda is scaled up if the ratio of the validation loss to the training loss
+    is above the tolerable interval, and scaled down if the ratio is below the
+    tolerable interval. Nothing happens if the ratio is within the tolerable
+    interval.
+    """
+
+    def __init__(self, scaling_factor: float, tolerable_interval: Tuple[float, float]):
+        """Initialize the interval parameter scaler.
+
+        Args:
+            scaling_factor: The factor by which to scale the lambda, a value in (0, 1).
+            tolerable_interval: The interval within which the ratio of the validation
+                loss to the training loss is considered acceptable. A tuple whose first
+                element is at least 0 and the second element is greater than the first.
+
+        Raises:
+            ValueError: If the tolerable interval is not a tuple of length 2.
+            ValueError: if the scaling factor is not in (0, 1).
+            ValueError: if the tolerable interval is negative or not a proper interval.
+        """
+        eps = np.finfo(float).eps
+        if not (eps < scaling_factor < 1 - eps):
+            raise ValueError(
+                "scaling_factor must be in (0, 1) within machine precision.",
+            )
+        if len(tolerable_interval) != 2:
+            raise ValueError("tolerable_interval must be a tuple of length 2")
+        if not (0 <= tolerable_interval[0] < tolerable_interval[1]):
```

Comment:
I think you mean "higher", that's what the ratio corresponds to, and you normally increase regularization when the training loss is low and validation loss is high.

I guess you have a point, but then you might intuitively also require tolerable_interval[0]` to be `<=1` since why would you reduce regularization if the training loss is lower than validation? However, this is what the original paper does -- they set the tolerable interval from 1.1 to 1.5 IIRC, meaning that when losses are approximately the same you bring regularization down. I guess I can imagine an algorithm that when this happens does the opposite, and it's not totally obvious to me why that would never make sense and therefore we should disallow it by design.

Reply:
Thanks, you're right I was interpreting the ratio in the opposite direction. I do find it odd that `tolerable_interval[0] > 1` -- I guess it's some kind of anti-hysteresis measure. Given this, fine to leave it as-is, we should be able to replicate DRLHP even if we don't necessarily agree with their parameter choices.